pkg/mcs/discovery/register.go

```diff
@@ -73,37 +73,43 @@ func (sr *ServiceRegister) Register() error {
 	go func() {
 		defer logutil.LogPanic()
 		for {
+		KeepaliveLoop:
 			select {
 			case <-sr.ctx.Done():
 				log.Info("exit register process", zap.String("key", sr.key))
 				return
 			case _, ok := <-kresp:
-				if !ok {
-					log.Error("keep alive failed", zap.String("key", sr.key))
-					// retry
-					t := time.NewTicker(time.Duration(sr.ttl) * time.Second / 2)
-					for {
-						select {
-						case <-sr.ctx.Done():
-							log.Info("exit register process", zap.String("key", sr.key))
-							t.Stop()
-							return
-						default:
-						}
-
-						<-t.C
+				if ok {
+					continue
+				}
+				log.Error("keep alive failed", zap.String("key", sr.key))
+				// retry
+				t := time.NewTicker(time.Duration(sr.ttl) * time.Second / 2)
+				for {
+					select {
+					case <-sr.ctx.Done():
+						log.Info("exit register process", zap.String("key", sr.key))
+						t.Stop()
+						return
+					case <-t.C:
 						resp, err := sr.cli.Grant(sr.ctx, sr.ttl)
 						if err != nil {
 							log.Error("grant lease failed", zap.String("key", sr.key), zap.Error(err))
-							t.Stop()
 							continue
 						}
 
 						if _, err := sr.cli.Put(sr.ctx, sr.key, sr.value, clientv3.WithLease(resp.ID)); err != nil {
 							log.Error("put the key failed", zap.String("key", sr.key), zap.Error(err))
-							t.Stop()
 							continue
 						}
+
+						kresp, err = sr.cli.KeepAlive(sr.ctx, resp.ID)
+						if err != nil {
+							log.Error("client keep alive failed", zap.String("key", sr.key), zap.Error(err))
+							continue
+						}
+						t.Stop()
```

Comment:
What about extracting a function returns `kresp`? It's kind of annoying to introduce a nested `for-select` block here.